### PR TITLE
Add range

### PR DESCRIPTION
**Pull request by pouya-eghbali**

### CODE REVIEW

File: packages/internals/src/flow.js [hunk 1]
```diff
@@ -21,7 +19,8 @@ class Flow {
     return new Flow(this.scope, data);
   }
   map(fn, ...args) {
-    const data = this.data.map(item => {
+    const map = fn.isLazy ? this.data.lazyMap : this.data.map;
+    const data = map.call(this.data, item => {
       let fun = fn instanceof Method ? fn.get(this.data) : fn;
       if (typeof fun == "function") fun = new JSFn(fun);
```
Comment (garritfra) [inline packages/internals/src/flow.js]: This should probably be `===`. If you want, we can turn #134 into a maintainance issue that takes care of multiple things, including this.

File: packages/internals/src/io.js [hunk 2]
```diff
@@ -15,7 +13,7 @@ class IO {
...
 const io = fn => new IO(fn);
 
-module.exports = {
-  io,
-  IO
-};
+module.exports.io = io;
+module.exports.IO = IO;
+
+const { Lazy } = require("./lazy");
```
Comment (garritfra) [inline packages/internals/src/io.js]: Is there a reason the import is at the bottom of the file?

File: packages/internals/src/lazy.js [hunk 3]
```diff
@@ -18,7 +16,7 @@ class Lazy {
...
 const lazy = fn => new Lazy(fn);
 
-module.exports = {
-  lazy,
-  Lazy
-};
+module.exports.lazy = lazy;
+module.exports.Lazy = Lazy;
+
+const { Flow } = require("./flow");
```
Comment (garritfra) [inline packages/internals/src/lazy.js]: Same here, import at the bottom?

File: packages/internals/src/lazyMap.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    return this.toArray().map(fn);
+  }
+  lazyMap(fn) {
+    return new LazyMap({ getter: i => this.get(i), length: this.length, fn });
+  }
+}
+
+module.exports.LazyMap = LazyMap;
+
+const { Array } = require("./array");
```
Comment (garritfra) [inline packages/internals/src/lazyMap.js]: Same here, import at the bottom?

File: packages/internals/src/range.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    const value = this.start + this.step * index;
+    if (this.step < 0 && value < this.end) throw "Index out of range";
+    if (this.step > 0 && value > this.end) throw "Index out of range";
+    return value;
+  }
+}
+
+module.exports.Range = Range;
+
+const { Array } = require("./array");
```
Comment (garritfra) [inline packages/internals/src/range.js]: Same here, import at the bottom?

File: packages/internals/src/flow.js (hunk 1 above)
Comment (pouya-eghbali) [reply packages/internals/src/flow.js]: Yeah, let's do that. We should probably have some helper functions like isObject, isFunction

File: packages/internals/src/io.js (hunk 2 above)
Comment (pouya-eghbali) [reply packages/internals/src/io.js]: This solves cyclic dependencies being undefined on import

**Comment by pouya-eghbali** [comment]
@garritfra I don't know how to solve that. We have 2 types of map. A lazy map, for mapping lazy functions to an array or range, and a regular map for mapping non-lazy functions (like print). The lazy map method returns a LazyMap instance, so for example if you do

```clio
[:] -> * double -> * add 1
```

It'll result in a LazyMap, and there'll be no `Allocation failed - JavaScript heap out of memory` errors. However, when you do `map`, an Array will be returned (After all, that's what map means, it returns a result). When map method is called, it creates an empty array, gets one element at a time from our data, passes it to the function, and appends the return value to this array, finally returning the array as its result. When the size of array becomes bigger than the amount of memory our process can use, that error happens.

**Comment by garritfra** [comment]
@pouya-eghbali I just tried the example you provided. For me, it results in an `Error: undefined`. Was this just an explanatory example?

**Comment by pouya-eghbali** [comment]
Yeah exactly, double and add aren't defined

**Comment by garritfra** [comment]
Oh, of course😂Sorry, my bad!

**Pull request merged**